x-pack/plugin/blob-cache/src/main/java/org/elasticsearch/blobcache/shared/SharedBlobCacheService.java

```diff
@@ -590,6 +592,128 @@ public void maybeFetchRegion(
+     * @param fetchExecutor an executor to use for reading from the blob store
+     * @param listener      a listener that is completed with {@code true} if the current thread triggered the fetching of the region, in
+     *                      which case the data is available in cache. The listener is completed with {@code false} in every other cases: if
+     *                      the region to write is already available in cache, if the region is pending fetching via another thread or if
+     *                      there is not enough free pages to fetch the region.
+     */
+    public void maybeFetchRegions(
+        final KeyType cacheKey,
+        final int firstRegion,
+        final int lastRegion,
+        final long blobLength,
+        final RangeMissingHandler writer,
+        final Executor fetchExecutor,
+        final ActionListener<Void> listener
+    ) {
+        int regionCount = lastRegion + 1 - firstRegion;
+        ArrayList<CacheFileRegion<KeyType>> regionsToFetch = new ArrayList<>(regionCount);
+        for (int i = firstRegion; i <= lastRegion; i++) {
+            if (freeRegionCount() < 1 && maybeEvictLeastUsed() == false) {
+                // no free page available and no old enough unused region to be evicted
+                logger.info("No free regions, skipping loading regions [{}-{}]", i, lastRegion);
+                break;
+            } else {
+                try {
+                    final CacheFileRegion<KeyType> entry = get(cacheKey, blobLength, i);
+                    entry.incRefEnsureOpen();
+                    regionsToFetch.add(entry);
+                } catch (AlreadyClosedException e) {
+                    // Fine to ignore. It is only added to the list after we have inc()
+                }
+
+            }
+        }
+        if (regionsToFetch.isEmpty()) {
+            listener.onResponse(null);
+            return;
+        }
+
+        try (
+            RefCountingListener regionsListener = new RefCountingListener(
+                ActionListener.releaseAfter(listener, () -> regionsToFetch.forEach(AbstractRefCounted::decRef))
+            )
+        ) {
+            try {
+                final List<Tuple<CacheFileRegion<KeyType>, RegionGaps>> gaps = new ArrayList<>(regionsToFetch.size());
+                for (CacheFileRegion<KeyType> toFetch : regionsToFetch) {
+                    int region = toFetch.regionKey.region();
+                    ByteRange regionRange = ByteRange.of(0, computeCacheFileRegionSize(blobLength, region));
+                    if (regionRange.isEmpty() == false) {
+                        List<SparseFileTracker.Gap> regionGaps = toFetch.tracker.waitForRange(
+                            regionRange,
+                            regionRange,
+                            regionsListener.acquire()
+                        );
+                        if (regionGaps.isEmpty() == false) {
```

Comment:
I wonder if we should break it up if we hit a region that needs no data read? We could in principle risk having everything but the first and last region of a file and thus we could download an excess amount of extra data.

We can also build this into the shared stream handler.

Or forego it for now. I'd be ok with that too, but wanted to get views on it.